upcoming: [M3-9109] - Update Firewall devices Linode landing table to account for new interfaces

coliu-akamai opened the pull request:
## Description 📝
- Adds Linode Interfaces to the firewall device Linode table

## Changes  🔄
-  Add Network Interface column to Linode device table
- Add query to fetch all Linodes that have an associated Linode Interface firewall device for preprocessing stuff
- Update copy as needed (ex: removing a Linode Interface device)

## Preview 📷

| Before  | After   |
| ------- | ------- |
| ![image](https://github.com/user-attachments/assets/d8f7a437-d461-4c9b-a67c-7376f775b7e6) | ![image](https://github.com/user-attachments/assets/901f24c9-886f-42a4-821a-bcf969b36df5) |

## How to test 🧪

### Prerequisites
- Using devenv or devcloud,

### Verification steps
With feature flag on,
- Confirm Linode Interface devices show up as expected in the device table only for Linodes table, not for Nodebalancer table
  - confirm Network Interface column
  - confirm label link takes you to Linode's page (for now)
- Confirm sorting by label still works as expected for devices in Firewall device table, even with Linode Interface devices
- Confirm text copies make sense when removing a Linode Interface device

With feature flag off,
- confirm Linode Interface related stuff does not appear (no extra column, rows for interface devices shouldn't show up)

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support

<br/>

- [ ] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [ ] All unit tests are passing
- [ ] TypeScript compilation succeeded without errors
- [ ] Code passes all linting rules

</details>

coliu-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceRow.tsx:
```diff
@@ -1,29 +1,64 @@
+  // we need the Linode ID to create a link
+  const entityId = isInterfaceDevice ? Number(url.split('/')[4]) : id;
+
+  // Since we're not given the Linode's label for a Linode Interface device, we must fetch it
+  const { data: linode, isLoading } = useLinodeQuery(
+    entityId,
+    isInterfaceDevice
+  );
```
Firewall device example for Linode Interfaces:
```
{
    "created": "2018-01-01T00:01:01",
    "entity": {
      "id": 123,
      "label": null,
      "type": "interface",
      "url": "/v4/linode/instances/123/interfaces/123"
    },
    "id": 456,
    "updated": "2018-01-02T00:01:01"
},

```
currently sending a request to get a Linode Interface's Linode. It has downsides:
- sending the request - if there are many interfaces assigned to the firewall, we'd send a bunch of requests
- sorting by entity label is unresponsive for Linode Interfaces (unless I fetch the relevant Linodes in `FirewallDeviceTable` and do some processing there?? >> maybe `useAllLinodesQuery` filtered by the Linode IDs we need, update any interface `device.entity.label` to make its label be the relevant label, etc)

cc @bnussman-akamai - is it worth asking the API to provide the Linode's label instead of returning 'null' for the `device.entity`'s label? Otherwise I can look into the preprocessing idea

coliu-akamai replied:
update: revisiting this, showcasing entity links for any Linode Interface firewall devices in FirewallsLanding will also require a bunch of processing. Having the Linode's label from the start will make it a lot easier and not require extra queries in a bunch of different places - will ask API about the possibility. If not, will look into what I can do

coliu-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceTable.tsx:
```diff
@@ -35,11 +44,56 @@ export const FirewallDeviceTable = React.memo(
+
+    // only fire this query if we have linode interface devices. We fetch the Linodes those devices are attached to
+    // so that we can add a label to the devices for sorting and display purposes
+    const { data: linodesWithInterfaces } = useAllLinodesQuery(
+      {},
+      filterForInterfaceDeviceLinodes,
+      linodeInterfaceDevices.length > 0
+    );
+
+    const updatedDevices = devices.map((device) => {
+      if (device.entity.type === 'interface') {
+        const linodeId = Number(device.entity.url.split('/')[4]);
+        const associatedLinode = linodesWithInterfaces?.find(
+          (linode) => linode.id === linodeId
+        );
+        return {
+          ...device,
+          entity: {
+            ...device.entity,
+            label: associatedLinode?.label ?? null,
+          },
+        };
+      } else {
+        return device;
+      }
+    });
```
Using a single query to fetch all Linodes with linode interface firewall entities instead of fetching each individual Linode per firewall table row - see previous comments for background

coliu-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceRow.tsx:
```diff
@@ -1,29 +1,59 @@
+  return (
+    <TableRow data-testid={`firewall-device-row-${id}`}>
+      <TableCell>
+        {/* The only time a firewall device's label comes in as null is for Linode Interface devices. This label won't stay null - we do some
```
Not sure how to get around this - I had the same issue when individually querying for the Linode, where I had to show a loading indicator (otherwise the label column is blank for a bit)

coliu-akamai reviewed (commented):
API may change the firewall device entity to include more information, so will potentially revisit the queries here

bnussman-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceTable.tsx:
```diff
@@ -35,11 +44,56 @@ export const FirewallDeviceTable = React.memo(
+          []
+        : [];
+    const filterForInterfaceDeviceLinodes: Filter = {
+      ['+or']: linodeInterfaceDevices.map((device) => {
+        return { id: Number(device.entity.url.split('/')[4]) };
+      }),
```
I like the idea here, and it would probably work for 99.99% of cases, but we have had to hotifx bugs in the past caused by large X-Filters. We may just want to use `useAllLinodesQuery` without any params/filters here to be safe. 

- https://github.com/linode/manager/pull/9151

bnussman-akamai commented on packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceRow.tsx:
```diff
@@ -1,29 +1,59 @@
+        {/* The only time a firewall device's label comes in as null is for Linode Interface devices. This label won't stay null - we do some
+        processing to give the interface device its associated Linode's label. However, processing may take time, so we show a loading indicator first */}
+        {isInterfaceDevice && !label ? (
+          <CircleProgress size="xs" />
```
Let's use a `Skeleton` loading state here

cpathipa commented on packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceRow.tsx:
```diff
@@ -1,29 +1,59 @@
+  const isInterfaceDevice = type === 'interface';
+  // for Linode Interfaces, the url comes in as '/v4/linode/instances/:linodeId/interfaces/:interfaceId
+  // we need the Linode ID to create a link
+  const entityId = isInterfaceDevice ? Number(url.split('/')[4]) : id;
```
Can we use `useParams` hook here ? 

example : 
```
const { linodeId } = useParams();
  const entityId = Number(linodeId);
```

cpathipa commented on packages/manager/src/features/Firewalls/FirewallDetail/Devices/FirewallDeviceTable.tsx:
```diff
@@ -35,11 +44,56 @@ export const FirewallDeviceTable = React.memo(
+
+    const updatedDevices = devices.map((device) => {
+      if (device.entity.type === 'interface') {
+        const linodeId = Number(device.entity.url.split('/')[4]);
```
Same here - useParams hook

cpathipa reviewed (approved):
Nice work @coliu-akamai confirming on the functionality.

coliu-akamai commented:
@bnussman-akamai - ty, addressed feedback!

@cpathipa - since we're in the Firewalls feature, we unfortunately can't get the Linode ID from params 😞 It is a bit janky, but will be changed soon - API has plans to update the entity object, which will make getting Linode IDs and labels for interface devices a lot cleaner on our end:

```
     {
            "id": 2,
            "created": "2025-01-09T16:04:09",
            "updated": "2025-01-09T16:04:09",
            "entity": {
                "id": 2,
                "type": "interface",
                "label": null,
                "url": "/v4/linode/instances/506951/interfaces/2",
            	   "parent_entity": {
                    "id": 506951,
                    "type": "linode",
             	       "label": "linode506951",
       	       "url": "/v4/linode/instances/506951",
"parent_entity": null
           }
            }
        },
```
